File: android/src/main/java/expo/modules/externaldisplay/ExpoExternalDisplayModule.kt
package expo.modules.externaldisplay

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoExternalDisplayModule : Module() {
  private var helper: ExternalDisplayHelper? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoExternalDisplay")

    Events(
      "@ExpoExternalDisplay_screenDidConnect",
      "@ExpoExternalDisplay_screenDidChange",
      "@ExpoExternalDisplay_screenDidDisconnect"
    )

    Function("getConstants") {
      mapOf(
        "SCREEN_INFO" to getScreenInfo()
      )
    }

    Function("init") {
      ensureHelper()
    }

    Function("getInitialScreens") {
      mapOf("SCREEN_INFO" to getScreenInfo())
    }

    View(ExpoExternalDisplayView::class) {
      Prop("screen") { view: ExpoExternalDisplayView, screen: String? ->
        view.setScreen(screen)
      }

      Prop("fallbackInMainScreen") { view: ExpoExternalDisplayView, fallback: Boolean? ->
        view.setFallbackInMainScreen(fallback == true)
      }

      GroupView {
        AddChildView { parent: ExpoExternalDisplayView, child: android.view.View, index: Int ->
          parent.addExternalDisplayChild(child, index)
        }

        GetChildCount { parent: ExpoExternalDisplayView ->
          parent.externalDisplayChildCount
        }

        GetChildViewAt<android.view.View> { parent: ExpoExternalDisplayView, index: Int ->
          parent.getExternalDisplayChildAt(index)
        }

        RemoveChildView { parent: ExpoExternalDisplayView, child: android.view.View ->
          parent.removeExternalDisplayChild(child)
        }

        RemoveChildViewAt { parent: ExpoExternalDisplayView, index: Int ->
          parent.removeExternalDisplayChildAt(index)
        }
      }

      OnViewDestroys { view: ExpoExternalDisplayView ->
        view.cleanup()
      }
    }

    OnDestroy {
      helper?.dispose()
      helper = null
    }
  }

  private fun ensureHelper(): ExternalDisplayHelper {
    val context = appContext.reactContext ?: appContext.currentActivity
    requireNotNull(context) { "No Android context available for ExpoExternalDisplay" }

    return helper ?: ExternalDisplayHelper(context.applicationContext, object : ExternalDisplayHelper.Listener {
      override fun onDisplayAdded() {
        sendEvent("@ExpoExternalDisplay_screenDidConnect", getScreenInfo())
      }

      override fun onDisplayChanged() {
        sendEvent("@ExpoExternalDisplay_screenDidChange", getScreenInfo())
      }

      override fun onDisplayRemoved() {
        sendEvent("@ExpoExternalDisplay_screenDidDisconnect", getScreenInfo())
      }
    }).also {
      helper = it
      ExpoExternalDisplayView.helperProvider = { it }
    }
  }

  private fun getScreenInfo(): Map<String, Any?> {
    return ensureHelper().getScreenInfo()
  }
}

File: ios/ExpoExternalDisplayModule.swift
import ExpoModulesCore
import UIKit

public class ExpoExternalDisplayModule: Module {
  private var didRegisterObservers = false

  public func definition() -> ModuleDefinition {
    Name("ExpoExternalDisplay")

    Events(
      "@ExpoExternalDisplay_screenDidConnect",
      "@ExpoExternalDisplay_screenDidChange",
      "@ExpoExternalDisplay_screenDidDisconnect"
    )

    Function("getConstants") {
      self.onMainActor {
        [
          "SCREEN_INFO": self.getScreenInfo()
        ] as [String: Any]
      }
    }

    Function("init") {
      self.onMainActor {
        self.registerObserversIfNeeded()
      }
    }

    Function("getInitialScreens") {
      self.onMainActor {
        ["SCREEN_INFO": self.getScreenInfo()]
      }
    }

    View(ExpoExternalDisplayView.self) {
      Prop("screen") { (view: ExpoExternalDisplayView, screen: String?) in
        view.screen = screen ?? ""
      }

      Prop("fallbackInMainScreen") { (view: ExpoExternalDisplayView, fallback: Bool?) in
        view.fallbackInMainScreen = fallback ?? false
      }
    }

    OnDestroy {
      NotificationCenter.default.removeObserver(self)
    }
  }

  private func onMainActor<T>(_ body: @MainActor () -> T) -> T {
    if Thread.isMainThread {
      return MainActor.assumeIsolated(body)
    }

    return DispatchQueue.main.sync {
      MainActor.assumeIsolated(body)
    }
  }

  @MainActor
  private func registerObserversIfNeeded() {
    if didRegisterObservers {
      return
    }
    didRegisterObservers = true

    let center = NotificationCenter.default
    center.addObserver(self, selector: #selector(handleScreenDidConnect), name: UIScreen.didConnectNotification, object: nil)
    center.addObserver(self, selector: #selector(handleScreenDidDisconnect), name: UIScreen.didDisconnectNotification, object: nil)
    center.addObserver(self, selector: #selector(handleScreenDidChange), name: UIDevice.orientationDidChangeNotification, object: nil)
  }

  @MainActor
  private func getScreenInfo() -> [String: Any] {
    var screenInfo: [String: Any] = [:]
    UIScreen.screens.enumerated().forEach { index, screen in
      guard screen != UIScreen.main else {
        return
      }
      screenInfo[String(index)] = [
        "id": index,
        "width": screen.bounds.width,
        "height": screen.bounds.height,
        "mirrored": screen.mirrored == UIScreen.main
      ]
    }
    return screenInfo
  }

  @objc private func handleScreenDidConnect() {
    onMainActor {
      sendEvent("@ExpoExternalDisplay_screenDidConnect", getScreenInfo())
    }
  }

  @objc private func handleScreenDidChange() {
    onMainActor {
      sendEvent("@ExpoExternalDisplay_screenDidChange", getScreenInfo())
    }
  }

  @objc private func handleScreenDidDisconnect() {
    onMainActor {
      sendEvent("@ExpoExternalDisplay_screenDidDisconnect", getScreenInfo())
    }
  }
}
